Fix #2630 : "Delete" label rendering on embedded collection + add space between checkbox and label when using icheck

vincenttouzet opened the pull request:
As asked here : https://github.com/sonata-project/SonataAdminBundle/issues/2630#issuecomment-196755129

* Add space between checkbox and label when using icheck
* Hide checkbox label on embedded collection tables

When using `sonata_type_collection` with `['edit' => 'inline', 'inline => 'table']`. The label of boolean fields are not visible (as they're already in the table heading). For expanded choices the label is still visible.
![capture](https://cloud.githubusercontent.com/assets/2057992/14043085/c840c432-f280-11e5-8a9b-d021393e3651.PNG)

vincenttouzet commented:
@OskarStark I'v fix tests failure + rebase and squash my commit on the latest master

ju1ius commented on Resources/views/Form/form_admin_fields.html.twig:
```diff
@@ -166,7 +166,9 @@ file that was distributed with this source code.
         <label{% for attrname, attrvalue in label_attr %} {{ attrname }}="{{ attrvalue }}"{% endfor %}>
             {{- widget|raw -}}
-            {{- label is not same as(false) ? label|trans({}, translation_domain) -}}
+            {%- if sonata_admin.options['use_icheck'] -%}&nbsp;{%- endif -%}
```
This should be fixed using css !

vincenttouzet replied:
I agree but in a previous pull request I fixed with CSS but @Soullivaneuh and @greg0ire said that 

> It feels less hacky that the css override

ju1ius replied:
Wow... :scream: 
Spacing with `&nbsp;` reminds me the old Dreamweaver days...

ju1ius replied:
```css
label > [class^="icheckbox_"],
label > [class^="iradio_"] {
    margin-right: 0.5rem;
}
```
see https://github.com/sonata-project/SonataAdminBundle/pull/3647/files#diff-56522bc6fbcd8161b013b8181f017c54R1

greg0ire replied:
@vincenttouzet : here is [the comment you are referring to](https://github.com/sonata-project/SonataAdminBundle/pull/3127#discussion_r54783538). It was about something else, and I personally prefer when the override can be done without CSS, because it seems less fragile (making an override in CSS implies relying on a class that may change for non-obvious reasons). In this case, I can you please follow @ju1ius advice? He's way better at front-end matters than I am.

ju1ius replied:
> making an override in CSS implies relying on a class that may change for non-obvious reasons

Yep, valid point.
I tend to think about non-breaking chars as a typographic feature (keeping two related words together on the same line)...
The advantage is that an override can be easily overriden :wink: !
And you can add a nice explainer in your stylesheet:
```css
/**
 * The iCheck checkboxes & radios have 0 margin by default,
 * add some space for the label text.
 */
label > [class^="icheckbox_"],
label > [class^="iradio_"] {
    margin-right: 0.5rem;
}
```

greg0ire replied:
>  I tend to think about non-breaking chars more as a typographic feature (keeping two related words together on the same line)...

Me too !

> The advantage is that a css override can be easily overriden :wink: !

Of course this is great, but you must also try to keep the link between CSS and the we are using `icheck` very obvious. Since CSS does not really have namespaces, a newcomer might not know iCheck and think this is something created within the project. This is why, in this other PR where you refactor CSS, you were kind of forced to create a `legacy.css` file containing things that might or might not still be in used, who knows?

In short, using CSS is great, but if you can avoid it without relying on hack like the `&nbsp;` hack, I would recommend avoiding it.

ju1ius replied:
yeah css quickly becomes a nightmare to maintain as soon as a project grows...
Things are moving a bit however. We'll soon have variables, maybe nesting...
In the meantime, tooling & conventions...

vincenttouzet replied:
Originaly I wrote this fix in `Resources/public/css/styles.css` (Inspired by bootstrap)
```css
.checkbox div[class^="icheckbox"],
.checkbox-inline div[class^="icheckbox"],
.radio div[class^="icheckbox"],
.radio-inline div[class^="icheckbox"] {
    position: relative;
    margin-top: 4px \9;
    margin-left: -20px;
    margin-right: 5px;
    margin-top: -3px;
}
.form-inline .checkbox div[class^="icheckbox"],
.form-inline .checkbox-inline div[class^="icheckbox"],
.form-inline .radio div[class^="icheckbox"],
.form-inline .radio-inline div[class^="icheckbox"] {
    position: relative;
    margin-top: 4px \9;
    margin-left: 0;
    margin-right: 5px;
    margin-top: -3px;
}
```

So I can remove the `use_icheck` options test in https://github.com/sonata-project/SonataAdminBundle/blob/master/Resources/views/Form/form_admin_fields.html.twig#L83 so the widget is rendered the same whether `use_icheck` option is `true`or `false`.

What do you think @ju1ius ?

ju1ius replied:
> so the widget is rendered the same whether use_icheck option is true or false.

I'm having trouble understanding what you mean by that. Mind a screenshot ?

ju1ius replied:
You mean you want to wrap the icheck element in a `div.checkbox` ?

vincenttouzet replied:
> You mean you want to wrap the icheck element in a div.checkbox ?

Yes that's it.

ju1ius replied:
For what pupose ?

vincenttouzet replied:
If the `label` is wrapper in a `div.checkbox` then the label won't be bold.

With just your CSS fix :

![css fix](https://cloud.githubusercontent.com/assets/2057992/13984332/c6a81d00-f0f5-11e5-9209-3078450b7fbe.PNG)


When wrapping with `div.checkbox` + my css fix : 

![bootstrap fix](https://cloud.githubusercontent.com/assets/2057992/13984344/d9c6cd96-f0f5-11e5-83f5-0a718b7a1a40.PNG)

vincenttouzet replied:
The purpose of wrapping with `div.checkbox` is to remove the check on `sonata_admin.options['use_icheck']` on `checkbox_widget` and `radio_widget` blocks

```twig
{% block checkbox_widget -%}
    {% set parent_label_class = parent_label_class|default('') -%}
    {% if 'checkbox-inline' in parent_label_class or sonata_admin.options['use_icheck'] %}
        {{- form_label(form, null, { widget: parent() }) -}}
    {% else -%}
        <div class="checkbox">
            {{- form_label(form, null, { widget: parent() }) -}}
        </div>
    {%- endif %}
{%- endblock checkbox_widget %}
```

Becomes : 

```twig
{% block checkbox_widget -%}
    {% set parent_label_class = parent_label_class|default('') -%}
    {% if 'checkbox-inline' in parent_label_class %}
        {{- form_label(form, null, { widget: parent() }) -}}
    {% else -%}
        <div class="checkbox">
            {{- form_label(form, null, { widget: parent() }) -}}
        </div>
    {%- endif %}
{%- endblock checkbox_widget %}
```

ju1ius commented:
> If the label is wrapped in a div.checkbox then the label won't be bold.

I do not think it is a good idea to make the checkbox/radio labels font-weight different from the other field types.
An example from a custom blog post admin:

* use_icheck: false

  ![use_icheck-false-fs8](https://cloud.githubusercontent.com/assets/218404/14006750/e0d35778-f170-11e5-954b-c1e46df749b2.png)

* use_icheck: true

  ![use_icheck-true-fs8](https://cloud.githubusercontent.com/assets/218404/14006772/ff8345d4-f170-11e5-98b5-9023df4b5473.png)

We can see that when use_icheck === false, the label rendering is inconsistant.
Moreover, in this case, the published and pinned fields are top-level boolean fields. They are important in the publication process, not just choices in a list.

Another example from SonataUserBundle:
* use_icheck: false
![user-icheck-false-fs8](https://cloud.githubusercontent.com/assets/218404/14006911/236e12f2-f172-11e5-9746-d95d15fb52a7.png)
* use_icheck: true
![user-icheck-true-fs8](https://cloud.githubusercontent.com/assets/218404/14006917/31e41cdc-f172-11e5-937b-daab1e3c00f2.png)

So, as for the labels font-weight, **the fix should be applied the other way around**, i.e. by doing:
```css
.checkbox label,
.radio label {
  font-weight: 700;
}
```

ju1ius commented:
> Remove "Delete" label rendering on embedded collection

Please, for the sake of accessibility, **never, ever, remove text from a form label**. Never.
If you must do so for presentational reasons, hide the label text with CSS, and do it in an accessible way (`display:none` verboten).
Please refer to [this article](http://webaim.org/techniques/forms/) on accessible forms ([this section](http://webaim.org/techniques/forms/advanced#invisible) is about hiding labels).

ju1ius commented:
IMO, the right direction would be to get rid of iCheck (which is currently unmaintained btw) and implement a one, true, CSS-only solution for styling checkboxes & radios, which is quite a simple thing to do [nowadays](http://caniuse.com/#feat=css-sel3).

vincenttouzet commented:
Ok @ju1ius  I've made the changes.

For the delete label I wrapped it in a `<span class="sr-only"></span>`.

I added your CSS to make labels on radios and checkboxes consistant with other labels (I removed the check on `use_icheck` options on `checkbox_widget` and `radio_widget` blocks so I removed the `Sonata\AdminBundle\Tests\Form\Widget\FormChoiceWidgetTest::testBootstrapLabelRendering` test.

But I don't understand the error on SonataCI. Maybe it comes from another commit (I rebased on the latest master)

greg0ire commented:
> But I don't understand the error on SonataCI. Maybe it comes from another commit (I rebased on the latest master)

The master branch build is passing, and Travis fails too…

jvasseur commented:
@greg0ire it is possible the master branch would fail if you run the test again on it.

It look like an update of phpunit or one of the dependency broke something. Maybe the same update that broke the symfony tests (https://github.com/sebastianbergmann/phpunit-mock-objects/issues/299)

greg0ire commented:
> @greg0ire it is possible the master branch would fail if you run the test again on it.

It is unlikely, but possible nonetheless. Looks like we need to tighten the version constraints then ?

greg0ire commented:
Ok, so from what I understand, we should stop configuring methods that do not exist or something like that. In the meantime, @vincenttouzet, can you force phpunit-mock-objects to < 3.1.2 ?

jvasseur commented:
The strange thing here is that the method `searchAndRenderBlock` exists and should be configurable.

greg0ire commented:
> The strange thing here is that the method searchAndRenderBlock exists and should be configurable.

Maybe it is not part of the thing being mocked (an interface ?)

greg0ire commented:
>         $formExtension = $this->getMock('Twig_ExtensionInterface', array('renderListElement', 'initRuntime', 'getTokenParsers', 'getNodeVisitors', 'getFilters', 'getTests', 'getFunctions', 'getOperators', 'getGlobals', 'getName'));

It is not part of the list of mocked methods.

jvasseur commented:
Ho, didn't see we were limiting the list of mocked methods. Look like this is the problem.

greg0ire commented:
Can either of you fix this in a separate PR (since it is kind of urgent) ? #makeTheBuildGreenAgain

jvasseur commented:
I can try.

greg0ire commented:
Thanks!

ju1ius commented:
@vincenttouzet Oh there's one more thing...
```css
.checkbox div[class^="icheckbox"]
```
You really don't need that `div` in your selectors.

```css
margin-top: 4px \9;
```
You can remove that hack. [Old IE if officially dead !](https://www.microsoft.com/en-us/WindowsForBusiness/End-of-IE-support) :tada:

ju1ius commented on Resources/views/Form/form_admin_fields.html.twig:
```diff
@@ -166,7 +166,14 @@ file that was distributed with this source code.
+                {%- else -%}
+                    {{ label|trans({}, translation_domain) }}
+                {%- endif -%}
+            {%- endif -%}
```
For consistency (and future-proofness), the following would be better.
```twig
{%- if label is not same as(false) -%}
  <span class="{{ name == '_delete' ? 'sr-only': '' }}">
    {{- label|trans({}, translation_domain) -}}
  </span>
{%- endif -%}
```

ju1ius replied:
Also it would be nice if this solution worked for all embedded checkboxes, not just `_delete`.
Although I don't really know if this doable...
Maybe with a custom function that walks up the form's ancestor and returns whether the current one is a child of `sonata_type_collection` ?
Anybody has an idea ?

ju1ius replied:
Ok it turns out to be possible:
```php
public function isEmbeddedInCollectionTable(FormView $form)
{
    while ($form = $form->parent) {
        if (empty($form->vars['sonata_admin']) || empty($form->vars['sonata_admin']['field_description'])) {
            break;
        }
        /** @var FieldDescriptionInterface $fd */
        $fd = $form->vars['sonata_admin']['field_description'];
        if ($fd->getType() === 'sonata_type_collection' && $fd->getOption('inline') === 'table') {
            return true;
        }
    }
    return false;
}
```

Of course this should apply only for single-checkbox field labels.

greg0ire commented:
Rebase on the latest master should fix the build.

ju1ius commented:
@vincenttouzet Also, your implementation removes the delete checkbox label **for all field types**, so that won't work.
For example inside a` sonata_type_collection` with `['edit' => 'inline', 'inline => 'natural']`, the checkbox would have no label **at all**...

The simplest thing I can think of right now is this:
```twig
{%- if label is not same as(false) -%}
  <span class="control-label__text">
    {{- label|trans({}, translation_domain) -}}
  </span>
{%- endif -%}
```
```css
.sonata-ba-field-inline-table td[class*="-_delete"] .control-label__text {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0,0,0,0);
  border: 0;
}
```

vincenttouzet commented:
Indeed. I think its it's the easier way to fix this. I update my PR

vincenttouzet commented:
If I change the selector with the following : 

```css
.sonata-ba-field-inline-table td > div.checkbox > label > .control-label__text
```

We can hide the label for each checkbox on the table (as the label is already in the table heading it's repeated. What's your opinion on this @ju1ius  ?

ju1ius commented:
@vincenttouzet Have you tested with fields like expanded choice widget (sonata_type_model with option expanded => true, or something like that) ?

vincenttouzet commented:
Yes I've tested and it works as the choices are wrapper in `ul` and `li`

vincenttouzet commented:
If it's ok for you I can rebase and squash my commits

OskarStark commented:
@vincenttouzet can you please post your conclusion in this long ticket and what is the final approach, thank you

@greg0ire can you please have a look?

vincenttouzet commented:
@OskarStark I've just rebase and squash + update the first post

OskarStark commented:
Thank you
